margarine/marginal_stats.py

```diff
@@ -130,43 +136,131 @@ def mask_arr(arr):
         de_logL, mask = mask_arr(logprob - base_logprob)
         dekl = np.average(de_logL, weights=self.theta_weights[mask])
-        debd = 2*np.cov(de_logL, aweights=self.theta_weights[mask])
+        debd = 2 * np.cov(de_logL, aweights=self.theta_weights[mask])
         theta_logL, mask = mask_arr(theta_logprob - theta_base_logprob)
         thetakl = np.average(theta_logL, weights=self.theta_weights[mask])
-        thetabd = 2*np.cov(theta_logL, aweights=self.theta_weights[mask])
+        thetabd = 2 * np.cov(theta_logL, aweights=self.theta_weights[mask])
 
         kl_array = np.sort([midkl, dekl, thetakl])
         bd_array = np.sort([midbd, debd, thetabd])
 
-        results_dict = {'Statistic': ['KL Divergence', 'BMD'],
-                        'Value': [kl_array[1], bd_array[1]],
-                        'Lower Bound': [kl_array[0], bd_array[0]],
-                        'Upper Bound': [kl_array[2], bd_array[2]]}
+        results_dict = {
+            "Statistic": ["KL Divergence", "BMD"],
+            "Value": [kl_array[1], bd_array[1]],
+            "Lower Bound": [kl_array[0], bd_array[0]],
+            "Upper Bound": [kl_array[2], bd_array[2]],
+        }
 
         results = pd.DataFrame(results_dict)
-        results.set_index('Statistic', inplace=True)
+        results.set_index("Statistic", inplace=True)
 
         return results
+
+    def integrate(self, loglikelihood, batch_size=1000, sample_size=10000):
+        """
+        Importance sampling integration of a likelihood function
+        """
+        xs = np.empty((sample_size, self.de.theta.numpy().shape[-1]))
+        ys = np.empty(sample_size)
+
+        n_todo = sample_size
+        trials = 0
+        xs = np.empty((sample_size, self.de.theta.numpy().shape[-1]))
+        # ps = np.empty((sample_size, target.ndim))
+        fs = np.empty(sample_size)
+        gs = np.empty(sample_size)
+        jac = np.empty(sample_size)
+
+        n_todo = sample_size
+        trials = 0
+        with tqdm(total=sample_size) as pbar:
+            while n_todo > 0:
+                x = self.de.sample(batch_size).numpy()
+                f = np.array(list(map(loglikelihood, x)))
+                g = self.de.log_prob(x).numpy()
+                in_bounds = np.logical_and(f != 0, ~np.isnan(g))
+                n_accept = x[in_bounds].shape[0]
+                if n_accept <= n_todo:
+                    xs[
+                        sample_size - n_todo : sample_size - n_todo + n_accept
+                    ] = x[in_bounds]
+                    fs[
+                        sample_size - n_todo : sample_size - n_todo + n_accept
+                    ] = f[in_bounds]
+                    gs[
+                        sample_size - n_todo : sample_size - n_todo + n_accept
+                    ] = g[in_bounds]
+                    jac[
+                        sample_size - n_todo : sample_size - n_todo + n_accept
+                    ] = self.prior_de.logpdf(x[in_bounds]) 
```

Comment:
I think the prior comes in by including this, here I have furnished the calculate object with the scipy stats object corresponding to the gaussian polychord prior I used.

It might be worth making this log_prob method from margarine in line with the scipy logpdf method so this could seamlessly use either a margarine de or the official de in the case of running from a typical prior?

Reply:
Yes I think this looks right... equivalent to

$Z \approx \frac{1}{N} \sum_{i=1} \frac{L(x)}{\tilde{L}(x)}\pi(x)$

In my previous comment. Can we rename the variable from `jac` to `prior`?

Can you add a catch that checks if prior_de is a margarine object and set `prior_logpdf` to `prior_de.log_prob` else if scipy object set `prior_logpdf` to `prior_de.logpdf`? The alternative is to redefine the `maf.log_prob` method to `maf.logpdf` but this is quite a significant change that users will need to be aware of. Also the `log_prob` notation is inline with tensorflow probability package.

Reply:
Done, decided to make the method take a prior_pdf function as arg (a function returning the logpdf), so either `scipy.stats.norm.logpdf`, `margarine.MAF.log_prob` or a user defined function if it is the typical polychord use case of a mixture of priors.